packages/circuit-ui/components/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,349 @@
+   */
+  onClose?: () => void;
+  /**
+   * Fired when the dialog starts to close.
+   * Use this to trigger animations before the dialog closes.
+   */
+  onCloseStart?: () => void;
+  /**
+   * Text label for the close button for screen readers.
+   * Important for accessibility.
+   */
+  closeButtonLabel?: string;
+  /**
+   * One or more [IETF BCP 47](https://en.wikipedia.org/wiki/IETF_language_tag)
+   * locale identifiers such as `'de-DE'` or `['GB', 'en-US']`.
+   * When passing an array, the first supported locale is used.
+   * Defaults to `navigator.language` in supported environments.
+   */
+  locale?: Locale;
+  /**
+   * Prevent users from closing the modal by clicking/tapping the overlay or
+   * pressing the escape key. Default `false`.
+   */
+  preventClose?: boolean;
+  /**
+   * Enables focusing a particular element in the dialog content and override default behavior
+   */
+  initialFocusRef?: RefObject<HTMLElement>;
+  /**
+   * a ReactNode or a function that returns the content of the modal dialog.
+   */
+  children?:
+    | ReactNode
+    | (({ onClose }: { onClose?: DialogProps['onClose'] }) => ReactNode);
+  /**
+   * The duration of the dialog animation in milliseconds.
+   * If you wish to animate the dialog, provide a value of the animation duration to enable the animation to complete before the dialog closes.
+   * default: 0
+   */
+  animationDuration?: number;
+  [key: DataAttribute]: string | undefined;
+}
+
+export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
+  (props, ref) => {
+    const {
+      open,
+      isModal = false,
+      children,
+      onClose,
+      closeButtonLabel,
+      className,
+      initialFocusRef,
+      preventClose,
+      animationDuration = 0,
+      onCloseStart,
+      ...rest
+    } = props;
+    const dialogRef = useRef<HTMLDialogElement>(null);
+    const openRef = useLatest<boolean>(open);
+    const lastFocusedElementRef = useRef<HTMLElement | null>(null);
+
+    // eslint-disable-next-line compat/compat
+    const hasNativeDialog = window.HTMLDialogElement !== undefined;
+
+    useScrollLock(open && isModal);
+
+    // set initial focus on the modal dialog content
+    useEffect(() => {
+      const dialogElement = dialogRef.current;
+      let timeoutId: NodeJS.Timeout;
+      if (open && dialogElement) {
+        timeoutId = setTimeout(() => {
+          if (initialFocusRef?.current) {
+            initialFocusRef?.current?.focus();
+          } else {
+            const firstFocusableElement = getFirstFocusableElement(
+              dialogElement,
+              isModal,
+            );
+            if (firstFocusableElement) {
+              firstFocusableElement.focus();
+            } else {
+              dialogElement.focus();
+            }
+          }
+        }, animationDuration);
+      }
+      return () => {
+        clearTimeout(timeoutId);
+      };
+    }, [open, initialFocusRef?.current, isModal, animationDuration]);
```

Comment:
Ref values should practically never be referenced in hook dependencies, as that would cause the effect to re-run more often than necessary. 

The purpose of refs is that they are stable objects whose value can change _without_ causing a re-render.

```suggestion
    }, [open, initialFocusRef, isModal, animationDuration]);
```